### PR TITLE
add focus mode

### PR DESCRIPTION
**Pull request by jtoy**
hit control+f and it goes into focus mode. could have used command f, but didnt want to block that frequently used combination. here is a video of it in action: https://cl.ly/413C2t0N2A2M

### CODE REVIEW

**Comment by roundhill** [comment]
Very cool. I think it looks little odd to lose the toolbar completely. What if it just hid the tag/note list?

@drw158 any thoughts?

**Comment by jtoy** [comment]
Yeah that could work, I’m not a Mac developer, so I’m slow at this stuff, but this is my most important app I use now.

**Comment by drw158** [comment]
I like the idea of calling it "Focus Mode" or something like that. I also think it's smart to first add it as a menu item or keyboard shortcut. in this issue we explored how to add it in the UI https://github.com/Automattic/simplenote-electron/pull/443

This wouldn't be a true focus mode, but I would keep the toolbar in there. Maybe you could fade out the icons in the mode, but show them full opacity when you hover over them.

**Comment by Gatada** [comment]
Hm, there are several apps I use that offer multiple "levels" of focus, that the user can cycle through with multiple taps on the same shortcut. I like that.

I do not mind at all to loose the entire interface apart from the text. That is not an uncommon behaviour. But maybe adding a button to to exit focus mode would be a good idea? Some might hit the shortcut by mistake, making it hard to reverse.

Introducing it for SimpleNote is great. But for me, the main issue with the current implementation (shown in the video) is easy to fix: the claustrophobic margin around the text area. I would prefer at least a 60 points inset added to all edges (or ideally a fixed max width of the text area, maybe limited to 800 points or something).

**Comment by jtoy** [comment]
hi all, so a few different ideas,  what are next steps?

**Comment by jtoy** [comment]
Is this what we are looking for? https://cl.ly/3c3E0t3Z1V3W

**Comment by Gatada** [comment]
Oh I like that, except for the text padding. Am I the only one feeling the text being squeezed at the top?

**Comment by drw158** [comment]
@jtoy That is about right, except the top margin of the window shouldn't change. The text shouldn't jump to the top. Also, I think the toolbar and icons should always be visible, but they can fade out a little bit in focus mode. Sorry, in my last comment, by "fade out" I meant reduce in opacity, not completely invisible.

**Comment by jtoy** [comment]
@drw158 Like this? https://cl.ly/0s1M0R2X1Y32

**Comment by Gatada** [comment]
That gets my vote! Looks great and seems to work as expected. Nice work.

**Comment by drw158** [comment]
That's much better! Can the toolbar/icons go up to 100% opacity on hover? Is the focus mode activated with a keyboard shortcut?

I think we could start testing and reviewing cc @roundhill @jleandroperez

**Comment by roundhill** [comment]
@jtoy can you please push your latest changes? I think the way you are adding the editor as a subview will cause memory leaks if you do it a bunch of times. Have you tried just collapsing all of the split controller layouts except for the editor? I think that'd be the proper way to do it.

Since you're not a Mac developer we can take it over and implement as we've designed if you'd like.

**Comment by jtoy** [comment]
@roundhill here is my new pull request, thank you!

**Comment by roundhill** [comment]
@jtoy your recent commits look much better 👍 I'll review soon.

File: Simplenote/SimplenoteAppDelegate.m [hunk 1]
```diff
@@ -230,6 +230,8 @@ - (void)configureToolbar
     
     splitFrame.size.height                      -= toolbarFrame.size.height;
     self.splitView.frame                        = splitFrame;
+    self.splitViewTopConstraint.constant = toolbarFrame.size.height;
```
Comment (roundhill) [inline Simplenote/SimplenoteAppDelegate.m]: This property doesn't exist when I build the app so it doesn't compile.

File: Simplenote/SimplenoteAppDelegate.m [hunk 2]
```diff
@@ -421,13 +423,13 @@ - (void)handleWindowDidBecomeMainNote:(NSNotification *)notification
...
     // Tag List: Don't draw separators
-    return (dividerIndex == SPSplitViewSectionTags);
+    return (dividerIndex == SPSplitViewSectionTags) || (dividerIndex == SPSplitViewSectionList);
 }
 
 - (NSRect)splitView:(NSSplitView *)splitView effectiveRect:(NSRect)proposedEffectiveRect forDrawnRect:(NSRect)drawnRect ofDividerAtIndex:(NSInteger)dividerIndex
 {
     // Tag List: Don't draw separators
-    return (dividerIndex == SPSplitViewSectionTags) ? CGRectZero : proposedEffectiveRect;
+    return (dividerIndex == SPSplitViewSectionTags) || (dividerIndex == SPSplitViewSectionList && self.noteListViewController.isCollapsed) ? CGRectZero : proposedEffectiveRect;
```
Comment (roundhill) [inline Simplenote/SimplenoteAppDelegate.m]: `noteListViewController.isCollapsed` doesn't exist. Did you forget to commit your changes to `NoteListViewController`?

File: Simplenote/SimplenoteAppDelegate.m [hunk 3]
```diff
@@ -660,17 +665,51 @@ - (void)searchAction:(id)sender
...
 
+- (IBAction)focusModeAction:(id)sender {
+    CGFloat tagListSplitPosition = MAX([self tagListSplitPosition], SPSplitViewDefaultWidth);
+    CGFloat editorSplitPosition = [self editorSplitPosition];
+    BOOL collapsed = !self.noteListViewController.isCollapsed;
+    self.noteListViewController.isCollapsed = collapsed;
+    [self.noteListViewController.searchField setHidden:collapsed];
+    [self.tagListViewController.view setHidden:YES];
+    if (collapsed) {
+//        self.splitViewTopConstraint.constant = 0;
```
Comment (roundhill) [inline Simplenote/SimplenoteAppDelegate.m]: Please remove any commented code!

File: Simplenote/SimplenoteAppDelegate.m [hunk 4]
```diff
@@ -660,17 +665,51 @@ - (void)searchAction:(id)sender
...
+        [self.splitView applyStyle];
+        [self.toolbar setAlphaValue:0.3];
+//        [self.toolbar setHidden:YES];
+//        self.splitView.toolbar = self.toolbar;
+    }
+    else {
```
Comment (roundhill) [inline Simplenote/SimplenoteAppDelegate.m]: We typically collapse these: `} else {`

**Comment by jtoy** [comment]
@roundhill just added the rest of the changes, had issues with my computers, hopefully this is all you need :)

File: Simplenote/SPSplitView.m [hunk 5]
```diff
@@ -150,4 +150,36 @@ - (void)applyStyle
...
+//    }
+//    else if (!self.toolbar.isHidden && (mY > self.toolbar.frame.size.height || mY < 0 || mX < 0 || mX > self.frame.size.width))  {
+//        [NSAnimationContext runAnimationGroup:^(NSAnimationContext * _Nonnull context) {
+//            context.duration = 0.5;
+//            self.toolbar.animator.alphaValue = 0.0;
+//        } completionHandler:^{
+//            [self.toolbar setHidden:YES];
+//        }];
+//    }
+//}
```
Comment (roundhill) [inline Simplenote/SPSplitView.m]: Why was all of this code commented out?

**Comment by roundhill** [comment]
Another bug, if you launch the app when you left it in focus mode, the toolbar buttons are not faded.

**Comment by roundhill** [comment]
I think we're heading in the right direction here. I really appreciate your effort @jtoy, but I think maybe it'd be best to have one of the Simplenote engineers do another PR to get the code looking right. Ping @jleandroperez, any thoughts on this?

**Comment by jtoy** [comment]
@roundhill agreed, thanks !

**Comment by jleandroperez** [comment]
Hey @roundhill and @jtoy !! sorry i'm late for the party, got caught during a trip!!.

Although the changes look great, i would suggest few things!!. [iA Writer](https://ia.net/writer/) has a focus mode similar to what @jtoy has in mind (probably!!). Key things to improve the UX are, probably:

- Left / Right / Top margins, so that this looks like a sheet of paper.
- This mode should _probably_ take the app into Fullscreen mode. While windowed, feels a bit broken. But in fullscreen mode, you get the Immersive groove.
- Whenever the user exits Fullscreen, this mode should probably get disabled.
- And of course, regular edition while in fullscreen should be still supported.

Thanks everyone!!!

**Comment by jtoy** [comment]
hi all, just checking in, what are next steps on this?

**Comment by roundhill** [comment]
@drw158 I think we could use some direction here. If you like the UI that is proposed here I think we can move forward. @jleandroperez has some alternative ideas on how it can work above.

Other users also just really want to hide the notes list, maybe that could be a logical first step.

**Comment by drw158** [comment]
I'll research what @jleandroperez describes above and report back.

**Comment by roundhill** [comment]
@drw158 Did you have a chance to have a look?

**Comment by jtoy** [comment]
Great, how do we proceed?

**Comment by roundhill** [comment]
I think I'll create another PR using this one as the inspiration, if that works for you @jtoy

**Comment by jtoy** [comment]
Sounds good to me :)

**Comment by jtoy** [comment]
hi all, just checking in :) is there a way I can help move this along more? this is such an awesome feature.

**Comment by roundhill** [comment]
> hi all, just checking in :) is there a way I can help move this along more? this is such an awesome feature.

I'm working on it in this branch: https://github.com/Automattic/simplenote-macos/tree/add/focus-mode. Should have the PR ready soon!

**Comment by roundhill** [comment]
Closing in favor of #229

**Pull request closed without merging**